src/main/java/org/truffleruby/core/module/ModuleOperations.java

```diff
@@ -397,32 +397,25 @@ public static ConstantLookupResult lookupConstantWithInherit(RubyContext context
     public static MethodLookupResult lookupMethodCached(DynamicObject module, String name,
             DeclarationContext declarationContext) {
         final ArrayList<Assumption> assumptions = new ArrayList<>();
+        DynamicObject[] refinements;
 
-        // Look in ancestors
         for (DynamicObject ancestor : Layouts.MODULE.getFields(module).ancestors()) {
-            final ModuleFields fields = Layouts.MODULE.getFields(ancestor);
-            assumptions.add(fields.getMethodsUnmodifiedAssumption());
-            final InternalMethod method = fields.getMethod(name);
-            if (method != null) {
-                if (method.isRefined()) {
-                    if (declarationContext != null) {
-                        final DynamicObject[] refinements = declarationContext.getRefinementsFor(ancestor);
-                        if (refinements != null) {
-                            for (DynamicObject refinement : refinements) {
-                                final MethodLookupResult refinedMethod = lookupMethodCached(refinement, name, null);
-                                if (refinedMethod.isDefined()) {
-                                    for (Assumption assumption : refinedMethod.getAssumptions()) {
-                                        assumptions.add(assumption);
-                                    }
-                                    return new MethodLookupResult(refinedMethod.getMethod(), toArray(assumptions));
-                                }
-                            }
+            if ((refinements = getRefinementsFor(declarationContext, ancestor)) != null) {
+                for (DynamicObject refinement : refinements) {
+                    final MethodLookupResult refinedMethod = lookupMethodCached(refinement, name, null);
```

Comment:
With C < SuperClass < Object, looking for method `foo`,
Is this correct? Won't it look in all of the refinement module (R)'s ancestors and so actually look in Object?
I guess it's fine because if the method is actually in SuperClass it will find it first.
But, since R will "copy" C.ancestors when defined, it might miss a later `C.include SomeModule`, which might have a `foo` method.
Could you check the behavior on MRI and with this PR?
I can also write an example if it's not clear.

Reply:
Actually, the `include` (or `prepend`) seems to be reflected in `R.ancestors` too on MRI:

```ruby
class Object
  def foo
    :Object
  end
end

class SuperClass
  def foo
    :superclass
  end
end

class MyClass < SuperClass
end

module M
  R = refine MyClass do
    def foo2
      :refinement
    end
  end
end

module SomeModule
  def foo
    :SomeModule
  end
end

using M

p MyClass.ancestors
p M::R.ancestors
# [MyClass, SuperClass, Object, Kernel, BasicObject]
# [#<refinement:MyClass@M>, MyClass, SuperClass, Object, Kernel, BasicObject]

# MyClass.include SomeModule
MyClass.prepend SomeModule

p MyClass.ancestors
p M::R.ancestors
# [SomeModule, MyClass, SuperClass, Object, Kernel, BasicObject]
# [#<refinement:MyClass@M>, SomeModule, MyClass, SuperClass, Object, Kernel, BasicObject]

p MyClass.new.foo # => :SomeModule
```

So I suspect MRI really sets R's "superclass" (`ModuleFields(R).parentModule`) to be `ModuleFields(C).start` for `R = refine(C)`.
I think `Module#include` itself doesn't do that but instead copies the whole chain (with iclass/IncludedModule which are basically module pointers):
```ruby
module Included
end

module IncludedLater
end

module IncludeMe
  include Included
end

class A
  include IncludeMe
end

IncludeMe.include IncludedLater

class B
  include IncludeMe
end

p A.ancestors
# [A, IncludeMe, Included, Object, Kernel, BasicObject]
p B.ancestors
# [B, IncludeMe, IncludedLater, Included, Object, Kernel, BasicObject]
```

Reply:
Good catch! I need to test it to say how to achieve the same R.ancestors with late include/prepend.

Reply:
I think `ModuleFields(R).parentModule = ModuleFields(C).start` should work when creating R, similar to how it's done in setSuperClass().
I wonder why R was not simply a Class in MRI, then it could be a proper subclass of the refined class: `R = refine C # => R = Class.new(C)`.
The problem is then that doesn't work if we're refining a module. So there is also some magic there where R has another module as "superclass/parent" and is affected by late include/prepend.

Reply:
> I think ModuleFields(R).parentModule = ModuleFields(C).start should work when creating R, similar to how it's done in setSuperClass().

https://github.com/oracle/truffleruby/pull/2033/commits/c9838085b7506a9843137a072aa6952572482e3f

The new test passes and all other too (except 3 MRI). 

> The problem is then that doesn't work if we're refining a module

```ruby
module A
  def foo?
    "foo from A"
  end
end

module IncludeMeLater
  def foo?
    "foo from IncludeMeLater"
  end
end

module C
  include A
end

refinement = Module.new do
  R = refine(C) do; end
end

puts R.ancestors
# (MRI) => [<refinement:C@#<Module:>>, C, BasicObjec]
# (TR)   => [<refinement:C@#<Module:>>, C, A]

result = nil
foo = Class.new do
  using refinement
  C.include(IncludeMeLater)
  include C

  puts C.ancestors
  # (MRI) => [C, IncludeMeLater, A]
  # (TR)   => [C, IncludeMeLater, A]

  puts R.ancestors # => (3)
  # (MRI) => [<refinement:C@#<Module:>>, C, BasicObjec]
  # (TR)   =>[<refinement:C@#<Module:>>, C, IncludeMeLater, A]
end

puts foo.new.foo? 
# (MRI) => "foo from IncludeMeLate"
# (TR)  =>  "foo from IncludeMeLate"
```

If we don't execute `C.include(IncludeMeLater)` both return "foo from A". It seems that MRI does not show all `R.ancestors`. But we are the same in terms of result.

Reply:
I'd even say that TR is more consistent, because we get the same behavior for refined classes and modules (in term of R.ancestors).
But then again - if the lookup is equal, maybe we can not pay attention to the different of R.ancestors.

Reply:
Looks OK to me.
A bit puzzling why MRI doesn't show `A` in the ancestors initially. Maybe a bug :smile: 
It's weird we don't see `BasicObject` on TruffleRuby though.

Maybe the logic to set parentModule should be more like it's done in `ModuleFields#initCopy`?

Reply:
Ah, but C is a module, so no BasicObject should be expected.

Reply:
> Ah, but C is a module, so no BasicObject should be expected.
Yeah, for me it looks like a bug to in MRI 🤷‍♀️